### PR TITLE
update kzgbn254 function to work with eigenDA encoded blobs

### PR DESCRIPTION
**Pull request by afkbyte**
works with blobs encoded by https://github.com/Layr-Labs/eigenda/pull/583

also uses some modified functions from the kzg library, diff can be found here:
https://github.com/Layr-Labs/rust-kzg-bn254/compare/master...afkbyte:rust-kzg-bn254:master?expand=1

when we recieve the data from the preimage it's already in evaluation formation not coefficient (has been ifftd before being dispersed) so we don't need to do a `g1_ifft` on the g1 SRS bc  kzgcommitment = <C,  IFFT(SRS)> = <IFFT(C), SRS>

### CODE REVIEW

File: arbitrator/prover/src/kzgbn254.rs [hunk 1]
```diff
@@ -169,3 +207,33 @@ fn append_left_padded_biguint_be(vec: &mut Vec<u8>, biguint: &BigUint) {
...
+    for elem in data.iter().rev() {
+        let bytes = elem.into_bigint().to_bytes_le(); 
+        buffer.write_all(&bytes);
+    }
+    // why is it like this?
+    buffer.reverse();
+}
+
+fn decode_codec_blob_header(codec_blob_header: &[u8]) -> (u8, u32){
+    let version = codec_blob_header[0];
```
Comment (ethenotethan) [inline arbitrator/prover/src/kzgbn254.rs]: lets do a length verification first so we can never encounter an index out of bounds exception

File: arbitrator/prover/src/kzgbn254.rs [hunk 2]
```diff
@@ -169,3 +207,33 @@ fn append_left_padded_biguint_be(vec: &mut Vec<u8>, biguint: &BigUint) {
...
+}
+
+// pls don't change this this is the correct deserialization omg
+pub fn serialize_montgomery_elements(data: &[Fr], buffer: &mut Vec<u8>) {
+    for elem in data.iter().rev() {
+        let bytes = elem.into_bigint().to_bytes_le(); 
+        buffer.write_all(&bytes);
+    }
+    // why is it like this?
+    buffer.reverse();
```
Comment (ethenotethan) [inline arbitrator/prover/src/kzgbn254.rs]: Do we have to reverse the buffer? If not mistaken we can refactor like so:
```
pub fn serialize_montgomery_elements(data: &[Fr], buffer: &mut Vec<u8>) {
    for elem in data.iter() {
        let bytes = elem.into_bigint().to_bytes_le(); 
        buffer.write_all(&bytes);
    }
    }

```

Comment (ethenotethan) [reply arbitrator/prover/src/kzgbn254.rs]: also use of for loops like this is typically non-idiomatic in rust - maybe lets use a for each pattern for dumping to the buffer:
```
data.iter().for_each(|x|{
    let b = x.into_bigint().to_bytes_le();
    buffer.write_all(&b);
})
```

File: arbitrator/prover/src/utils.rs [hunk 3]
```diff
@@ -299,23 +301,31 @@ pub fn hash_preimage(preimage: &[u8], ty: PreimageType) -> Result<[u8; 32]> {
...
-            let unpadded_preimage = unpadded_preimage_vec.as_slice();
+            
 
             // repad it here, TODO: need to ask to change the interface for this
-            let blob = EigenDABlob::from_bytes_and_pad(unpadded_preimage);
+            let blob = EigenDABlob::new(preimage.to_vec(), true);
             
-            let commitment = kzg_bn254.blob_to_kzg_commitment(&blob).unwrap();
+            let blob_polynomial = blob.to_polynomial().unwrap();
+            let commitment = kzg_bn254.commit_to_evaluation_polynomial(&blob_polynomial).unwrap();
```
Comment (ethenotethan) [inline arbitrator/prover/src/utils.rs]: lets do more graceful error handling via match versus struct unwrapping

**Comment by teddyknox** [comment]
Awesome work @afkbyte. Can we rebase and clean up comments etc? Would love to get this merged.

File: arbitrator/prover/src/kzgbn254.rs (hunk 2 above)
Comment (afkbyte) [reply arbitrator/prover/src/kzgbn254.rs]: modified into iterators!

Comment (afkbyte) [reply arbitrator/prover/src/kzgbn254.rs]: we unfortunately do need to reverse the buffer

File: arbitrator/prover/src/utils.rs (hunk 3 above)
Comment (afkbyte) [reply arbitrator/prover/src/utils.rs]: https://github.com/Layr-Labs/nitro-private/pull/2/commits/2fe18856e023e470b93bdb3bb06413c22d54526a

File: arbitrator/prover/src/kzgbn254.rs [hunk 4]
```diff
@@ -45,21 +46,23 @@ pub fn prove_kzg_preimage_bn254(
...
-    // expand the roots of unity, should work as long as it's longer than chunk length and chunks
-    // from my understanding the data_setup_mins pads both min_chunk_len and min_num_chunks to 
-    // the next power of 2 so we can load a max of 2048 from the test values here
-    // then we can take the roots of unity we actually need (len polynomial) and pass them in
-    // @anup, this is a really gross way to do this, pls tell better way
-    kzg.data_setup_mins(1, 2048)?;
+    // expand roots of unity
+    let roots_of_unity = kzg.calculate_roots_of_unity(preimage.len() as u64);
+    match roots_of_unity {
+        Ok(roots_of_unity) => println!("Roots of unity: {:?}", roots_of_unity),
```
Comment (anupsv) [inline arbitrator/prover/src/kzgbn254.rs]: Probably need to remove this print ?

File: arbitrator/prover/src/kzgbn254.rs [hunk 5]
```diff
@@ -80,48 +83,71 @@ pub fn prove_kzg_preimage_bn254(
...
+        Err(err) => return Err(err.into()),
+    };
+
+    let blob_coefficients = blob_polynomial_coefficient_form.to_vec();
+    let mut blob_bytes = Vec::new();
+    deserialize_montgomery_elements(&blob_coefficients, &mut blob_bytes);
+
+    // blob header is the first 32 bytes of the blob bytes
+    let blob_header = blob_bytes[..32].to_vec();
+    println!("blob header {:?}", blob_header);
```
Comment (anupsv) [inline arbitrator/prover/src/kzgbn254.rs]: another one.

Comment (afkbyte) [reply arbitrator/prover/src/kzgbn254.rs]: https://github.com/Layr-Labs/nitro-private/pull/2/commits/c99c50df24b4a8c588048977f47e422d2820ccfd

File: arbitrator/prover/src/kzgbn254.rs (hunk 4 above)
Comment (afkbyte) [reply arbitrator/prover/src/kzgbn254.rs]: https://github.com/Layr-Labs/nitro-private/pull/2/commits/c99c50df24b4a8c588048977f47e422d2820ccfd

File: arbitrator/prover/src/kzgbn254.rs [hunk 6]
```diff
@@ -167,3 +194,27 @@ fn append_left_padded_biguint_be(vec: &mut Vec<u8>, biguint: &BigUint) {
...
 
+pub fn deserialize_montgomery_elements(data: &[Fr], buffer: &mut Vec<u8>) {
+    let mut temp_buffer: Vec<u8> = data.iter()
+        .rev()
+        .flat_map(|elem| elem.into_bigint().to_bytes_le())
+        .collect();
+    
+    temp_buffer.reverse();
+    buffer.extend(temp_buffer);
+}
```
Comment (anupsv) [inline arbitrator/prover/src/kzgbn254.rs]: Doesn't `to_byte_array` from the lib provide something similar or is this something more specific than that ?

File: arbitrator/prover/src/utils.rs [hunk 7]
```diff
@@ -297,26 +297,18 @@ pub fn hash_preimage(preimage: &[u8], ty: PreimageType) -> Result<[u8; 32]> {
                 3000
             ).unwrap();
```
Comment (anupsv) [inline arbitrator/prover/src/utils.rs]: You might want to use Prod numbers for these. In the lib, there is mainnet-data folder under tests. You can use those files to specify the input files.

File: arbitrator/prover/src/kzgbn254.rs (hunk 6 above)
Comment (afkbyte) [reply arbitrator/prover/src/kzgbn254.rs]: so it's a bit different, `to_byte_array` deserializes them in big endian but I need them in little endian

File: arbitrator/prover/src/kzgbn254.rs [hunk 8]
```diff
@@ -45,21 +46,23 @@ pub fn prove_kzg_preimage_bn254(
...
-    // the next power of 2 so we can load a max of 2048 from the test values here
-    // then we can take the roots of unity we actually need (len polynomial) and pass them in
-    // @anup, this is a really gross way to do this, pls tell better way
-    kzg.data_setup_mins(1, 2048)?;
+    // expand roots of unity
+    let roots_of_unity = kzg.calculate_roots_of_unity(preimage.len() as u64);
+    match roots_of_unity {
+        Ok(_) => (),
+        Err(err) => return Err(err.into()),
+    };
```
Comment (ethenotethan) [inline arbitrator/prover/src/kzgbn254.rs]:
```suggestion
    let roots_of_unity = kzg.calculate_roots_of_unity(preimage.len() as u64)?;
```

Or if you need to wrap the error:
```suggestion
    let roots_of_unity = kzg.calculate_roots_of_unity(preimage.len() as u64).map_err(|e| Err(e))?;
```

File: arbitrator/prover/src/kzgbn254.rs [hunk 9]
```diff
@@ -45,21 +46,23 @@ pub fn prove_kzg_preimage_bn254(
...
+    // preimage is already padded, unpadding and repadding already padded data can destroy context post IFFT
+    // as some elements in the bn254 field are represented by 32 bytes, we know that the preimage is padded
+    // to 32 bytes per DA spec as the preimage is retrieved from DA, so we can use this unchecked function
+    let blob = Blob::from_padded_bytes_unchecked(preimage);
 
-    // we are expecting the preimage to be unpadded when turned into a blob function so need to unpad it first
-    let unpadded_preimage_vec: Vec<u8> = remove_empty_byte_from_padded_bytes(preimage);
-    let unpadded_preimage = unpadded_preimage_vec.as_slice();
+    let blob_polynomial_evaluation_form = blob.to_polynomial(PolynomialFormat::InEvaluationForm).unwrap();
+    let blob_commitment = kzg.commit(&blob_polynomial_evaluation_form).unwrap();
```
Comment (ethenotethan) [inline arbitrator/prover/src/kzgbn254.rs]:
```suggestion
    let blob_polynomial_evaluation_form = blob.to_polynomial(PolynomialFormat::InEvaluationForm)?;
    let blob_commitment = kzg.commit(&blob_polynomial_evaluation_form)?;
```

File: arbitrator/prover/src/kzgbn254.rs [hunk 10]
```diff
@@ -80,48 +83,70 @@ pub fn prove_kzg_preimage_bn254(
...
     );
 
-    let offset_usize = usize::try_from(offset)?;
-    let mut proving_offset = offset;
+    // transform polynomial into coefficient form
+    let mut blob_polynomial_coefficient_form = blob_polynomial_evaluation_form.clone();
+    match blob_polynomial_coefficient_form.transform_to_form(PolynomialFormat::InCoefficientForm) {
+        Ok(_) => (),
+        Err(err) => return Err(err.into()),
+    };
```
Comment (ethenotethan) [inline arbitrator/prover/src/kzgbn254.rs]:
```suggestion

blob_polynomial_coefficient_form.transform_to_form(PolynomialFormat::InCoefficientForm)?;
```

File: arbitrator/prover/src/kzgbn254.rs [hunk 11]
```diff
@@ -80,48 +83,70 @@ pub fn prove_kzg_preimage_bn254(
...
+    deserialize_montgomery_elements(&blob_coefficients, &mut blob_bytes);
+
+    // blob header is the first 32 bytes of the blob bytes
+    let blob_header = blob_bytes[..32].to_vec();
+
+    // decode blob header, version is currently unused however in the future we probabky
+    let length = match decode_codec_blob_header(&blob_header) {
+        Ok((_, length)) => length,
+        Err(err) => return Err(err),
+    };
```
Comment (ethenotethan) [inline arbitrator/prover/src/kzgbn254.rs]:
```suggestion
         let (_, length) = decode_codec_blob_header(&blob_header)?;
```

File: arbitrator/prover/src/kzgbn254.rs [hunk 12]
```diff
@@ -80,48 +83,70 @@ pub fn prove_kzg_preimage_bn254(
...
+    let proven_y_fr = match blob_polynomial_coefficient_form.get_at_index(proving_offset as usize) {
+        Some(value) => value,
+        None => return Err(eyre::eyre!("Index out of bounds")),
+    };
 
-    let proven_y_fr = to_fr_array(&proven_y);
+    let z_fr = match kzg.get_nth_root_of_unity(proving_offset as usize) {
+        Some(value) => value,
+        None => return Err(eyre::eyre!("Failed to get nth root of unity")),
+    };
```
Comment (ethenotethan) [inline arbitrator/prover/src/kzgbn254.rs]:
```suggestion
let z_fr = kzg.get_nth_root_of_unity(proving_offset as usize)
    .ok_or_else(|| eyre::eyre!("Failed to get nth root of unity"))?;
```

File: arbitrator/prover/src/kzgbn254.rs [hunk 13]
```diff
@@ -80,48 +83,70 @@ pub fn prove_kzg_preimage_bn254(
...
     padded_proving_offset_bytes[32 - proving_offset_bytes.len()..].copy_from_slice(&proving_offset_bytes);
 
-    // in production we will first need to perform an IFFT on the blob data to get the expected y value
-    let mut proven_y = blob.get_blob_data();
-    let offset_usize = offset as usize; // Convert offset to usize
-    proven_y = proven_y[offset_usize..(offset_usize + 32)].to_vec();
+    let proven_y_fr = match blob_polynomial_coefficient_form.get_at_index(proving_offset as usize) {
+        Some(value) => value,
+        None => return Err(eyre::eyre!("Index out of bounds")),
+    };
```
Comment (ethenotethan) [inline arbitrator/prover/src/kzgbn254.rs]:
```suggestion
let proven_y_fr = blob_polynomial_coefficient_form.get_at_index(proving_offset as usize)
    .ok_or_else(|| eyre::eyre!("Index out of bounds"))?;
```

File: scripts/create-test-preimages.py [hunk 14]
```diff
@@ -28,18 +26,29 @@ def kzg_test_data():
...
-        bytes_32[0] = 0
-        data.extend(bytes_32)
+    #     # 0 padding for 1st byte of 32 byte word
+    #     bytes_32[0] = 0
+    #     data.extend(bytes_32)
 
-    print(bytes(data))
-    return bytes(data)
+    # print(data)
+    # return bytes(data)
```
Comment (ethenotethan) [inline scripts/create-test-preimages.py]: lets remove this

File: arbitrator/prover/src/utils.rs [hunk 15]
```diff
@@ -297,26 +297,18 @@ pub fn hash_preimage(preimage: &[u8], ty: PreimageType) -> Result<[u8; 32]> {
...
-            let blob = EigenDABlob::from_bytes_and_pad(unpadded_preimage);
+            let blob = EigenDABlob::from_padded_bytes_unchecked(preimage);
             
-            let commitment = kzg_bn254.blob_to_kzg_commitment(&blob).unwrap();
+            let blob_polynomial = blob.to_polynomial(PolynomialFormat::InEvaluationForm).unwrap();
+            let blob_commitment = kzg_bn254.commit(&blob_polynomial).unwrap();
 
             let mut commitment_bytes = Vec::new();
-            commitment.serialize_uncompressed(&mut commitment_bytes).unwrap();
+            blob_commitment.serialize_uncompressed(&mut commitment_bytes).unwrap();
```
Comment (ethenotethan) [inline arbitrator/prover/src/utils.rs]: Let's avoid unwrapping whenever possible

**Review by ethenotethan — COMMENTED** [review]
mostly stylistic concerns around using idiomatic rust - otherwise builds and arbitrator tests are failing which should ideally be passing before merging

**Comment by ethenotethan** [comment]
Also looks like we're introducing cargo.lock dependency changes which are causing jit compilations to fail in CI - lets fix that too before merging

https://github.com/Layr-Labs/nitro-private/actions/runs/9370515667/job/25797665060?pr=2

**Comment by afkbyte** [comment]
> mostly stylistic concerns around using idiomatic rust - otherwise builds and arbitrator tests are failing which should ideally be passing before merging

stylistic issues addressed in https://github.com/Layr-Labs/nitro-private/pull/2/commits/ece9f226d31703f45522f0212f6f83fea56a4d80 and https://github.com/Layr-Labs/nitro-private/pull/2/commits/028802a58a3bf6b9ad9167dbedd415576c239375

python script comment removed in https://github.com/Layr-Labs/nitro-private/pull/2/commits/c505a1e10342a355eb6c06a681a3e9aa5ec38980

**Review by ethenotethan — APPROVED** [review]
LGTM

**Pull request merged**